File: backend/src/controllers/books.controller.ts
import { Request, Response } from 'express'
import { supabase } from '../config/supabase'

// GET /api/books
// Retorna todos os livros do usuário, do mais recente ao mais antigo
export const listarLivros = async (req: Request, res: Response) => {
  const userId = req.user!.id

  const { data, error } = await supabase
    .from('books')
    .select('*')
    .eq('user_id', userId)
    .order('created_at', { ascending: false })

  if (error) return res.status(400).json({ error: error.message })

  return res.status(200).json({ books: data ?? [] })
}

// POST /api/books
// Cria um novo livro. Body: { title, author, rating, review?, genres?, colorIdx? }
export const criarLivro = async (req: Request, res: Response) => {
  const userId = req.user!.id
  const { title, author, rating, review, genres, colorIdx } = req.body

  if (!title?.trim()) return res.status(400).json({ error: 'O título é obrigatório' })
  if (!author?.trim()) return res.status(400).json({ error: 'O autor é obrigatório' })

  const { data, error } = await supabase
    .from('books')
    .insert({
      user_id:   userId,
      title:     title.trim(),
      author:    author.trim(),
      rating:    rating   ?? 5,
      review:    review   ?? '',
      genres:    genres   ?? [],
      color_idx: colorIdx ?? 0,
    })
    .select()
    .single()

  if (error) return res.status(400).json({ error: error.message })

  return res.status(201).json({ book: data })
}

// DELETE /api/books/:id
// Remove o livro do usuário
export const deletarLivro = async (req: Request, res: Response) => {
  const userId = req.user!.id
  const bookId = req.params.id

  const { error } = await supabase
    .from('books')
    .delete()
    .eq('id', bookId)
    .eq('user_id', userId)

  if (error) return res.status(400).json({ error: error.message })

  return res.status(200).json({ message: 'Livro removido com sucesso' })
}

File: backend-go/internal/books/books_controller.go
package books

import (
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"beeplanerofc/backend-go/internal/dbutil"
	"beeplanerofc/backend-go/internal/middleware"
)

type Controller struct {
	db *pgxpool.Pool
}

func NewController(db *pgxpool.Pool) *Controller {
	return &Controller{db: db}
}

// GET /api/books — lista os livros do usuário, do mais recente ao mais antigo
func (ctrl *Controller) ListarLivros(c *gin.Context) {
	user := middleware.CurrentUser(c)

	rows, err := ctrl.db.Query(c.Request.Context(),
		"SELECT * FROM books WHERE user_id = $1 ORDER BY created_at DESC", user.ID,
	)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	livros, err := pgx.CollectRows(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"books": livros})
}

type criarLivroRequest struct {
	Title    string   `json:"title"`
	Author   string   `json:"author"`
	Rating   *int     `json:"rating"`
	Review   *string  `json:"review"`
	Genres   []string `json:"genres"`
	ColorIdx *int     `json:"colorIdx"`
}

// POST /api/books — cria um novo livro
func (ctrl *Controller) CriarLivro(c *gin.Context) {
	user := middleware.CurrentUser(c)

	var req criarLivroRequest
	c.ShouldBindJSON(&req)

	title := strings.TrimSpace(req.Title)
	author := strings.TrimSpace(req.Author)
	if title == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "O título é obrigatório"})
		return
	}
	if author == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "O autor é obrigatório"})
		return
	}

	rating := 5
	if req.Rating != nil {
		rating = *req.Rating
	}
	review := ""
	if req.Review != nil {
		review = *req.Review
	}
	genres := req.Genres
	if genres == nil {
		genres = []string{}
	}
	colorIdx := 0
	if req.ColorIdx != nil {
		colorIdx = *req.ColorIdx
	}

	rows, err := ctrl.db.Query(c.Request.Context(), `
		INSERT INTO books (user_id, title, author, rating, review, genres, color_idx)
		VALUES ($1, $2, $3, $4, $5, $6, $7)
		RETURNING *`,
		user.ID, title, author, rating, review, genres, colorIdx,
	)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	livro, err := pgx.CollectOneRow(rows, dbutil.RowToMap)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, gin.H{"book": livro})
}

// DELETE /api/books/:id — remove o livro do usuário
func (ctrl *Controller) DeletarLivro(c *gin.Context) {
	user := middleware.CurrentUser(c)
	id := c.Param("id")

	_, err := ctrl.db.Exec(c.Request.Context(),
		"DELETE FROM books WHERE id = $1 AND user_id = $2", id, user.ID,
	)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Livro removido com sucesso"})
}
